lib/trino-orc/src/main/java/io/trino/orc/reader/UUIDColumnReader.java

```diff
@@ -0,0 +1,266 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.orc.reader;
+
+import io.airlift.units.DataSize;
+import io.trino.orc.OrcColumn;
+import io.trino.orc.OrcCorruptionException;
+import io.trino.orc.metadata.ColumnEncoding;
+import io.trino.orc.metadata.ColumnMetadata;
+import io.trino.orc.stream.BooleanInputStream;
+import io.trino.orc.stream.ByteArrayInputStream;
+import io.trino.orc.stream.InputStreamSource;
+import io.trino.orc.stream.InputStreamSources;
+import io.trino.spi.TrinoException;
+import io.trino.spi.block.Block;
+import io.trino.spi.block.Int128ArrayBlock;
+import io.trino.spi.block.RunLengthEncodedBlock;
+import org.openjdk.jol.info.ClassLayout;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.VarHandle;
+import java.nio.ByteOrder;
+import java.time.ZoneId;
+import java.util.Optional;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static io.airlift.units.DataSize.Unit.GIGABYTE;
+import static io.trino.orc.metadata.Stream.StreamKind.DATA;
+import static io.trino.orc.metadata.Stream.StreamKind.PRESENT;
+import static io.trino.orc.stream.MissingInputStreamSource.missingStreamSource;
+import static io.trino.spi.StandardErrorCode.GENERIC_INTERNAL_ERROR;
+import static java.lang.Math.toIntExact;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class UUIDColumnReader
+        implements ColumnReader
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(UUIDColumnReader.class).instanceSize();
+    private static final int ONE_GIGABYTE = toIntExact(DataSize.of(1, GIGABYTE).toBytes());
+
+    private static final VarHandle LONG_ARRAY_HANDLE = MethodHandles.byteArrayViewVarHandle(long[].class, ByteOrder.nativeOrder());
```

Comment:
Parquet explicitly states bytes order depending on the type. I am pretty sure ORC does that too. Let's check and make it proper here, regardless of whether there is an incosistency somewhere else

Reply:
UUID is not supported in ORC and in v1 spec (https://orc.apache.org/specification/ORCv1/) I see some things encoded in little-endian and some in big-endian.

Reply:
The question is how popular writers handle that. We should just do what others are doing if the spec is not clear

Reply:
the new implementation is consistent with the old implementation using varbinary + all contemporary architectures use little-endian so this is not urgent to improve.
I created https://github.com/trinodb/trino/issues/13480

Reply:
Per https://datatracker.ietf.org/doc/html/rfc4122, the canonical representation for variant-1 UUIDs uses network byte order (i.e., big endian representation). We need to pick *some* format and stick to it -- ideally, otherwise, the encoding is not portable.

Regardless, using 64-bit little-endian (since we're treating values long-at-a-time) is not going to result in little endian for the whole 128-bit stream. For example, for a UUID with the following big-endian representation:

```
00 11 22 33 44 55 66 77 88 99 aa bb cc dd ee ff
```

the "little endian" interpretation from this  reader if `ByteOrder.nativeOrder()` is little endian would be:

```
77 66 55 44 33 22 11 00 ff ee dd cc bb aa 99 88
```

Reply:
My understanding is this (correct me if I'm wrong).
`java.util.UUID` java values are stored in two longs. On little-endian machines, those long values are stored in memory using little-endian so not conforming to the spec.
Trino stores uuid in Block as big-endian. It has to do a conversion but it doesn't check for native endianness so it assumes little-endian (see `io.trino.spi.type.UuidType#getObjectValue` or `io.trino.spi.type.UuidType#trinoUuidToJavaUuid`).
If the Block is `VariableWidthBlock` then the values can be stored directly in big-endian but if it's `Int128ArrayBlock`, those values are stored as two longs with reversed bytes vs the original java.util.UUID values.

Trino stores UUID values in the orc as BINARY and keeps the block byte order which means the values are stored as big-endian in orc (conforming to the spec).

Orc reader (the current and the new implementation) reads the long values using native byte order so little-endian thus it reads the same reversed long values. When the trino uuid value is presented is converted back to little-endian (bytes are reversed back).

For example, let's say we start with `java.utill.UUID (1L, 2L).`
in little-endian those values are `10000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000   01000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000`.
Stored in the `Int128ArrayBlock` are reversed values `00000000 00000000 00000000 00000000 00000000 00000000 00000000 10000000   00000000 00000000 00000000 00000000 00000000 00000000 00000000 02000000` (`2^56 2^57` in LE), the same values are stored in ORC.
When reading back to `Int128ArrayBlock`, we read two longs in LE so `2^56, 2^57`, when presenting this value to a user, trino converts it back to `java.utill.UUID (1L, 2L)` by reversing byte order.
Notice that long values are consistently kept in the "big-endian" order (the first long is the most significant bits and the second one is the least significant bits) but the bytes inside the longs are reordered.

Reply:
> java.util.UUID java values are stored in two longs. On little-endian machines, those long values are stored in memory using little-endian so not conforming to the spec.

That's irrelevant. What matters is how they get converted to/from a stream of bytes.

The two things that matter are:
* The on-disk encoding in ORC files. Values need to be portable across versions, query engines and architectures.
* The "block" representation within Trino. This is part of the contract between engine and connectors.

> If the Block is VariableWidthBlock then the values can be stored directly in big-endian but if it's Int128ArrayBlock, those values are stored as two longs with reversed bytes vs the original java.util.UUID values.

As long as the calls to `block.getLong()` return the same thing regardless of which block is being used and that we can describe/explain how those values relate to the bytes in the raw UUID canonical representation, it doesn't matter how each block stores things internally.

> Trino stores uuid in Block as big-endian. It has to do a conversion but it doesn't check for native endianness so it assumes little-endian

That's ok. There is an expectation that Trino runs on a little-endian machine.

> Trino stores UUID values in the orc as BINARY and keeps the block byte order which means the values are stored as big-endian in orc (conforming to the spec).

That's good.

Reply:
Ok, so to sum up.
Both UUID reader implementations, using `varbinary` and the new `UUIDColumnReader` return correct values. This all depends on the machine endianness being little-endian. I could change the `UUIDColumnReader` to always read in little endian but trino would be broken with regard to UUID on a big-endian machine anyway due to `UUIDType` value conversion.

Reply:
This whole conversation is completely off track. This is about UUID in Iceberg ORC files, so the only thing that matters [the Iceberg specification for UUID values](https://iceberg.apache.org/spec/#appendix-b-32-bit-hash-requirements):

> UUIDs are encoded using big endian. The test UUID for the example above is: `f79c3e09-677c-4bbd-a479-3f349cb785e7`. This UUID encoded as a byte array is: `F7 9C 3E 09 67 7C 4B BD A4 79 3F 34 9C B7 85 E7

Also, there is more than an expectation that Trino runs on little endian machines. Trino will [refuse to start](https://github.com/trinodb/trino/blob/71e11c35cb9342db67bc4bf60b00575224148c62/core/trino-main/src/main/java/io/trino/server/TrinoSystemRequirements.java#L65) if not little endian and [so will Slice](https://github.com/airlift/slice/blob/db02631a74de268be1dbe02d65352c5fa3d8ef13/src/main/java/io/airlift/slice/JvmUtils.java#L38).